elastic/kibana
# Index pattern field editor

This PR contains the reusable, embeddable everywhere, index pattern field editor. Currently, it is integrated into the index pattern management app. In future releases, it will be integrated into Discover, Lens, and any other app

The editor allows the user to

* Create and edit runtime fields
* Edit mapped fields (change the custom label, the format, or the popularity)

## Examples

The examples below assume that you have added the "indexPatternFieldEditor" plugin to the required plugins list (in kibana.json). You then get back its API through the start contract.

### Create / edit a field

```js
const MyComponent = () => {
    const closeFieldEditor = useRef<() => void | undefined>();

    // Access the plugin through context
    const { indexPatternFieldEditor } = useAppDependencies();

    const openEditor = () => {
        closeFieldEditor.current = indexPatternFieldEditor.openEditor({
            // the only required option is the context in which the editor is being used
            ctx: {
                indexPattern,
            },
            // optionally you can also provide the following parameters:
            fieldName: 'fieldToEdit',
            onSave: (field) => {
                // handler to call once the field has been saved 
                // e.g. refreshFieldList();
            },
        });
    };

    useEffect(() => {
        return () => {
            // Make sure to close the editor when unmounting
            if (closeFieldEditor.current) {
                closeFieldEditor.current();
            }
        }
    }, []);

    return (
        <EuiButton onClick={openEditor}>Create field</EuiButton>
    );
}
```

### Delete one or multiple fields

To delete one or multiple runtime fields there is a render props component that you can use.

```js

// The `deleteFields()` handler that you get back lets you pass a single (`string`)
// or multiple (`string[]`) field names to delete.

const MyComponent = () => {
    const { indexPatternFieldEditor } = useAppDependencies();
    const { DeleteRuntimeFieldProvider } = indexPatternFieldEditor;

    return (
        <DeleteRuntimeFieldProvider indexPattern={indexPattern}>
            {(deleteFields) => (
                <EuiButton fill color="danger" onClick={() => deleteFields('myField')}>
                    Delete
                </EuiButton>
            )}
        </DeleteRuntimeFieldProvider>       
    );
}
```

## How to test

* Navigate to the Stack management > Index pattern
* Click on one of your index patterns
* There should be a button to "Create field"
* Click on the button and start creating a field. The only required form field is the "name", al the others are optional.
* The painless script syntax is validated by executing a query against ES. You can try to pass invalid syntax and verify that you are not allowed to save the field.
* Once you save the field, it should appear in the table, along with any other field.
* You can now go to Discover and start using the field any way you want!

## API

The complete API is documented in the README.md

## Screenshots

<img width="1320" alt="Screenshot 2021-02-11 at 12 54 05" src="https://user-images.githubusercontent.com/2854616/107639202-77f6e580-6c68-11eb-8e9e-7166cd87ed68.png">
<img width="611" alt="Screenshot 2021-02-11 at 12 54 57" src="https://user-images.githubusercontent.com/2854616/107639211-7af1d600-6c68-11eb-8aa0-5136d296c27d.png">
<img width="818" alt="Screenshot 2021-02-11 at 12 55 27" src="https://user-images.githubusercontent.com/2854616/107639215-7b8a6c80-6c68-11eb-9660-23dced0da120.png">


## Release note

There is a new field editor in the index pattern management app that lets you create (and edit) runtime fields, adding them to the index pattern field list.
